crates/driver/src/blockchain/mod.rs

```diff
@@ -44,10 +47,33 @@ impl Ethereum {
         owner: eth::Address,
         spender: eth::allowance::Spender,
     ) -> Result<eth::allowance::Existing, Error> {
-        let amount = contracts::ERC20::at(&self.0, spender.token.0)
+        let amount = contracts::ERC20::at(&self.web3, spender.token.0)
             .allowance(owner.0, spender.address.0)
             .call()
             .await?;
         Ok(eth::Allowance { spender, amount }.into())
     }
 }
+
+// TODO This could probably do some caching, I might want to open an issue for
+// this?
+pub struct Contracts<'a>(&'a Ethereum);
```

Comment:
It would be nice if the methods that return the contract addresses weren't async. The contract address is determined by the chain id and hard coded into the contract deployment info. There is no network request involved there.

Reply:
Oh yeah! That is a nice idea, we can have `Contracts(Ethereum, Network)` and then the contract accessors would be sync and not require a `Result`:
```rust
let address = contracts::GPv2Settlement::raw_contract().get(&self.1).unwrap().address;
contracts::GPv2Settlement::at(&self.0.web3, address)
```